### ISSUE TITLE
Emitting a message to another namespace upon a connection event

### ISSUE THREAD

**Issue by yoavnash** [miguelgrinberg/Flask-SocketIO#1338]
Hi,

Thanks a lot for the great work done on this library!

As for my question, it seems that it's not possible to emit a message during the 'connect' event to another namespace.
I'm following the example here: https://flask-socketio.readthedocs.io/en/latest/#connection-events

Initially I used this code on the server-side:
``` python
@socketio.on('connect')
def on_connect():
    print('Client connected')
    flask_socketio.emit('my event', {'data': 'Connected'})
```

On the client-side I use the 'socket.io-client' package:
``` javascript
this.socket = io('/' + this.userId);
this.socket.on('connect', () => {
    console.log('Connected via websocket.');
});
this.socket.on('my event', (data) => {
    console.log('Got it:', data);
});
```

Even though I did see the print statement of the server-side (that is: `Client connected`), I couldn't see the data emitted on the client-side, as this screenshot of my browser's console shows:
![image](https://user-images.githubusercontent.com/62147380/88559320-f7a2c700-d02c-11ea-8b59-48cfb2b43330.png)

However, when I added the namespace to the function's decorator on the server-side, then everything worked: 
``` python
@socketio.on('connect', namespace='/9')
    def on_connect():
        print('Client connected')
        flask_socketio.emit('my event', {'data': 'Connected'})
```
The number 9 used for the namespace is the user ID of the current client.
Screenshot of the browser's console:
![image](https://user-images.githubusercontent.com/62147380/88560719-c62afb00-d02e-11ea-8cad-9f5c769fe44a.png)

My question is then, is it possible to emit a message upon a connection event to another namespace?
In other words, is it possible not to specify the namespace (as in my first attempt), but still be able to emit a message to a specific namespace?

The following variations didn't work for me:
``` python
@socketio.on('connect')
def on_connect():
    print('Client connected')
    flask_socketio.emit('my event', {'data': 'Connected'}, namespace='/9')
```

``` python
@socketio.on('connect')
def on_connect():
    print('Client connected')
    socketio.emit('my event', {'data': 'Connected'}, namespace='/9')
```

Shouldn't they work?

**Comment by miguelgrinberg** [comment]
This is what happens when a client connects to the '/9' namespace:

- a connection to the `/` namespace (the default namespace) is made. Emitting to `/9` at this point will not work because the `/9` namespace hasn't been created yet.
- a connection to `/9` is made after the connection to `/` is accepted.

If you only intend to work on the `/9` namespace, then you need to add `namespace='/9'` to all your handlers. Any handlers that don't have a `namespace` argument will be attached to the default namespace.

As a side note, you can't have variable namespaces, so `/{user_id}` is not a good use case for namespaces. I suggest you look at rooms instead.

**Comment by yoavnash** [comment]
If I understand correctly, then there is no way to emit a message to the client upon a connection event if the client chooses to connect with a namespace. Is that right?

**Comment by miguelgrinberg** [comment]
You can emit a message back to the client on any namespace they connect to. Just handle the `connect` event in the intended namespace. The client connects on the default namespace first, then on the custom namespace, so you cannot respond to the custom namespace from a default namespace handler, simply because at the time the default namespace is connected the client hasn't yet connected to the other namespace.

What is the issue that you have with multiple tabs? If you want all the tabs to share data, use a session cookie (i.e. Flask's `session` variable for example). If you want all the tabs to have independent data, use the `sid` as a user identifier, since it'll be different in each tab.

**Comment by yoavnash** [comment]
Ahh, now I see what you meant earlier - there are two connection events (one for `/` and a second one for `/9`). Now I see why the namespace approach doesn't fit - I'm mostly interested in the `/9` but namespaces in decorators can't be defined as `@socketio.on('connect', namespace='/<user_id>')` which is what I need in this case.

For the tabs, I also now realized that it's simpler than what I had in mind. A session cookie is one solution, but I think using the user ID as the room name is what works for me best.

Here's what I ended up with:
Server:
``` python
@socketio.on('connect')
def on_connect():
    print('Client connected')
    join_room(str(current_user.id))
    emit('my event', {'data': 'Connected'}, room=str(current_user.id))

@socketio.on('disconnect')
def on_disconnect():
    print('Client disconnected')
    leave_room(str(current_user.id))
```

Client:
``` javascript
this.socket = io();
this.socket.on('connect', () => {
    console.log('Connected via websocket.');
});
this.socket.on('my event', (data) => {
    console.log('Got it:', data);
});
```

Thanks for your help!
